substrate/client/network/src/discovery.rs

```diff
@@ -898,28 +923,55 @@ impl NetworkBehaviour for DiscoveryBehaviour {
 					},
 					KademliaEvent::OutboundQueryProgressed {
 						result: QueryResult::GetClosestPeers(res),
+						stats,
 						..
-					} => match res {
-						Err(GetClosestPeersError::Timeout { key, peers }) => {
+					} => {
+						let (key, peers, timeout) = match res {
+							Ok(GetClosestPeersOk { key, peers }) => (key, peers, false),
+							Err(GetClosestPeersError::Timeout { key, peers }) => (key, peers, true),
+						};
+
+						let target = match PeerId::from_bytes(&key.clone()) {
```

Comment:
dq: In litep2p we have a `key.preimage()` to fetch back the peer ID, does libp2p expose a similar API?
My understanding was that the key is a sha256 of the peer ID, and the conversion happens unidirectionally. In other words, is this the same peer ID?

Reply:
This is a different `key`, which can or cannot be a `PeerId`, internally stored as bytes. litep2p only supports looking up closest peers to keys which are `PeerId`s, but libp2p is more generic and uses arbitrary binary keys. To match litep2p backend I limited libp2p backend to `PeerId` keys only as well, but we still need to convert them back from the binary representation.